### ISSUE TITLE
Very slow build time to a NextJS app with only 2 pages

### ISSUE THREAD

**Issue by willianjusten** [aws-amplify/amplify-console#2127]
### Before opening, please confirm:

- [X] I have checked to see if my question is addressed in the [FAQ](https://github.com/aws-amplify/amplify-console/blob/master/FAQ.md).
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-console/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-console/blob/master/CONTRIBUTING.md).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### App Id

d1ywuiyssz1l16

### Region

us-east-2

### Amplify Console feature

Build settings, Performance

### Describe the bug

If you can see in the following image:

![image](https://user-images.githubusercontent.com/3991845/127967483-f480c2f8-b49a-4dea-ba26-e27c01ca6d14.png)

My builds are taking almost 10 minutes, but it's a very small project (NextJS with 2 static pages for now). I even removed testing stuff to run inside my CI/CD on Gitlab to save some time, but it still too slow... =/

Another thing that I noticed is that initially Amplify identified my project as Web Cypress even when I defined nextjs stuff.

---

If you could take a look, maybe it's just a misconfiguration, but it looks really weird.

### Expected behavior

It shouldn't take more than 2-3min, follow bellow the same project on Netlify and it's very similar in Vercel as well.

![image](https://user-images.githubusercontent.com/3991845/127967717-27810725-fd66-4a41-a689-1c967fbad9fe.png)

### Reproduction steps

1) Init the project with `amplify init` and answer the questions (remember it's next ssr project, so the build folder is .next)
2) Add hosting with `amplify add hosting`
3) Connect with Gitlab
4) Wait and see the build

### Build Settings

```yaml
version: 1
frontend:
  phases:
    preBuild:
      commands:
        - yarn
    build:
      commands:
        - yarn build
  artifacts:
    baseDirectory: .next
    files:
      - '**/*'
  cache:
    paths:
      - node_modules/**/*
```

**Comment by quinnturner** [comment]
+1, our organization is concerned with the build times as well.

The long build time is related to this application being an SSR application (instead of SSG). SSR applications currently take several minutes longer (my estimate is ~7.5 minutes). Changing to SSG will drastically improve build times but it has other bugs though: https://github.com/aws-amplify/amplify-console/issues/2085, so proceed with caution.

With SSR apps, I believe Amplify builds the application twice, once regularly and once for lambdas with `mode: "node"` on the Webpack config (completely unconfirmed, I am not an AWS employee). These long build times should be a resolvable issue on their end considering Vercel has a similar architecture under the hood and they are not plagued with these long builds. However, they might involve pretty significant changes in their build architecture to address the build times.

**Comment by willianjusten** [comment]
Oh, that's sad =\

Both Vercel and Netlify takes only 2min at maximum...

**Comment by medelman17** [comment]
Just wanted to comment that this remains an outstanding issue--have been dealing with excessively long build times just today.

<img width="540" alt="Screen Shot 2021-09-16 at 12 47 13 PM" src="https://user-images.githubusercontent.com/14793389/133654805-0342677f-de34-4dc8-9bf6-4c91ddce80cb.png">

At the very least, can the team provide any clarity with respect to what exactly is going on during the five+ minutes between: (1) SSR build Starting; and (2) SSR complete?

**Comment by quinnturner** [comment]
From @ferdingler on #2272:

> The reason for deployment times taking several minutes (~6 to 13 minutes) is because of the propagation time of the CloudFront distribution across all edge locations globally. The deployment itself (updating the code and creating resources) is actually pretty fast, but the rest of the time is just waiting for propagation to finish.
>
> We are aware that customers like faster deploy times and we are looking into ways to improve this. Thanks again!

**Comment by reedshea** [comment]
I'm experiencing a similar issue, I think - but it seems that although my application is purely a static site (SSG), Amplify still builds a server-side rendered version. The back-end CloudFormation work is pretty quick (3 minutes), and the front-end SSG build looks like it takes just a few seconds for this site. But the subsequent SSR build takes 7 minutes:

```
2021-11-28T21:19:06.524Z [INFO]: info  - Compiled successfully
2021-11-28T21:19:06.526Z [INFO]: info  - Collecting page data...
2021-11-28T21:19:08.270Z [INFO]: info  - Generating static pages (0/9)
2021-11-28T21:19:09.928Z [INFO]: info  - Generating static pages (2/9)
2021-11-28T21:19:09.936Z [INFO]: info  - Generating static pages (4/9)
2021-11-28T21:19:09.971Z [INFO]: info  - Generating static pages (6/9)
2021-11-28T21:19:10.093Z [INFO]: info  - Generating static pages (9/9)
2021-11-28T21:19:10.093Z [INFO]: info  - Finalizing page optimization...
2021-11-28T21:19:10.097Z [INFO]: 
[...]
2021-11-28T21:19:10.106Z [INFO]: λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
                                 ○  (Static)  automatically rendered as static HTML (uses no initial props)
                                 ●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
                                 (ISR)     incremental static regeneration (uses revalidate in getStaticProps)
2021-11-28T21:19:10.292Z [INFO]: Starting SSR Build...
2021-11-28T21:26:15.502Z [INFO]: SSR Build Complete.
2021-11-28T21:26:16.796Z [INFO]: # Completed phase: build
2021-11-28T21:26:16.821Z [INFO]: ## Build completed successfully
```

**Comment by quinnturner** [comment]
@reedshea make sure your `build` script in `package.json`:

```json
{
  "scripts": {
    "build": "next build && next export"
  }
}
```

The behaviour you're describing happens when the `build` script is just `next build`. When Amplify sees `next export`, if will forcibly only deploy the static assets and none of the Lambdas.

**Comment by reedshea** [comment]
Thank you, @quinnturner! I'll investigate and see whether that approach makes sense. There may be a need for SSR pages in my application, so I'll tread carefully and try to avoid running into hard-to-debug problems...

**Comment by nickmcmillan** [comment]
Same situation here. Small (10-ish SSR pages) NextJS site. No idea what's happening during these ~8 minutes. 

```
2021-12-03T04:05:18.299Z [INFO]: Starting SSR Build...
2021-12-03T04:13:09.007Z [INFO]: SSR Build Complete.
```

**Comment by manelephant** [comment]
I kind of don't understand how a small company like Vercel can implement a smooth deployment of a nextjs app in any version and amazon can't.

Wasted a full day on amplify trying to deploy nextjs 11 without success. Cryptic or false error reporting during build.

**Comment by llogan382** [comment]
> 

@quinnturner that recommendation worked great. These results are substantial: 

<img width="782" alt="Screen Shot 2021-12-14 at 1 22 43 PM" src="https://user-images.githubusercontent.com/26100101/146057500-763a692b-31a2-415d-b412-1622c0b38152.png">

**Comment by nickmcmillan** [comment]
I think that's fine if all that's needed is a static site - which is what `next export` generates. 
But if we want a site with any SSR pages - ie those that run on Next's nodejs environment - then we have to go through Amplify's very slow SSR build step.

**Comment by jayfry1077** [comment]
My site with 5 small pages and 1 hello world api took 17 minutes to build.

**Comment by entest-hai** [comment]
No excuse CloudFront, it should be much faster

**Comment by ivancsicsmarkus** [comment]
Many applications need SSR with faster build times. Our build times are around 12 minutes. More than 60% of that is "Starting SSR Build...". We understand that Cloudfront and Lambda@Edge need some time to update but it could be faster for such tasks where only a small number of assets and functions have to be updated.

**Comment by yauri-io** [comment]
I'm having the same issue. I'm looking for a different provider for hosting my NextJS project (SSR).
The build time on Amplify is too long, and I experienced 2 times stuck on build process, I just cancelled it.

**Comment by justyn-clark** [comment]
Same here and I'm ready to throw my laptop out of the window!!! 

I'm testing an issue only happening in Amplify and it takes 15mins each time for only a few pages. 

This is very painful.

**Comment by matart15** [comment]
1. created a new `nextjs` app.
2. just in case deleted `pages/api`. only 2 pages. `index.js` and `404`
3. `next build`.  Not `next build && next export` because we will use both SSG and SSR in the future.


|local|amplify on aws|
|---|---|
|time yarn build| phase.build.commands|
| 10.41s|12 minutes 39 seconds|

**Comment by PixellUp** [comment]
A very simple Next JS website took a whooping 40 min to deploy. This is unacceptable!!!

![image](https://user-images.githubusercontent.com/4704339/153211130-51e9a009-1d3d-4994-8497-82651addf4bd.png)

**Comment by justyn-clark** [comment]
> A very simple Next JS website took a whooping 40 min to deploy. This is unacceptable!!!
> 
> ![image](https://user-images.githubusercontent.com/4704339/153211130-51e9a009-1d3d-4994-8497-82651addf4bd.png)
> 
> Before
> 
> ![image](https://user-images.githubusercontent.com/4704339/153211396-cba0ee08-80ff-428e-9094-9befe1f45695.png)
> 

Dude I'm done with this setup and possibly Next altogether after this. Do yourself a favor lol ike I did and join the Remix.run team. It will change your life.

**Comment by justyn-clark** [comment]
> A very simple Next JS website took a whooping 40 min to deploy. This is unacceptable!!!
> 
> ![image](https://user-images.githubusercontent.com/4704339/153211130-51e9a009-1d3d-4994-8497-82651addf4bd.png)
> 
> Before
> 
> ![image](https://user-images.githubusercontent.com/4704339/153211396-cba0ee08-80ff-428e-9094-9befe1f45695.png)
> 

Dude I'm done with this setup and possibly Next JS and Gatsby altogether after this. 

Do yourself a favor like I did and join the Remix.run team and deploy on Fly.io. It will change your life.

**Comment by nickmcmillan** [comment]
[justyn-clark](https://github.com/justyn-clark) The problem this thread is discussing relates to Amplify's slow SSR build system, the bottleneck has nothing to do with using NextJS, even less to do with Gatsby.

**Comment by ajest** [comment]
Same problem here, we need to a solution

![Captura de Pantalla 2022-02-17 a la(s) 16 12 29](https://user-images.githubusercontent.com/4982093/154553671-4d1606c7-872a-4607-aefc-7ece2790d5f7.png)

**Comment by TomJD** [comment]
How is this issue open for such a long time and no fixes have been implemented?
Just as @manelephant said, how can Vercel achieve this working flawlessly and Amazon is still scratching its head?

**Comment by arshkkk** [comment]
Considering some caveats while working with AWS amplify for NextJS,  after that getting longer time ( so higher prices )

I moved to vercel, so much easy and hassle free

**Comment by calavera** [comment]
hey folks. We're working on improving this experience in our product. Unfortunately, it's not an easy fix, but we're taking steps in the right direction.

**Comment by TomJD** [comment]
Totally agree with @CauanCabral 

That comment will allow us to patiently wait until the next riot. Thank you for the feedback!

**Comment by m1stermanager** [comment]
@calavera is there any sort of ETA? if this lands prior to my app growing too large, i may be able to use amplify and just suffer through slow builds in the meantime.

**Comment by calavera** [comment]
@m1stermanager unfortunately, I cannot give you an ETA, sorry about that.

**Comment by leye195** [comment]
Hope it will be fixed soon... it take so long time on SSR build

**Comment by TimmDay** [comment]
@calavera Thanks for the update!
Don't know if this is possible - but maybe a temporary salve could be to provide a config option to turn on/off the cloudfront part of the amplify service, (or at least restrict it to edge locations in the local region where the s3 buckets are), so that build related debugging can happen faster?

**Comment by aungbobotun** [comment]
> The long build time is related to this application being an SSR application (instead of SSG). SSR applications currently take several minutes longer (my estimate is ~6-10 minutes) than SSG applications.
> 
> This long deployment time issue is likely not readily resolvable on Amplify's end. Vercel has a fundamentally different architecture for CDNs/Compute@Edge under the hood than AWS Amplify. Vercel uses Fastly (which is essentially Varnish-as-a-Service), whereas AWS uses CloudFront. Unlike Fastly, CloudFront does not support instant invalidations and propagations in CDNs & deployments of Compute@Edge. As a result, Amplify is bottlenecked by CloudFront.
> 
> I suggest updating the title of this post to include `SSR` since this is only related to SSR apps.

Even my local machine with RAM8GB just took Max:2min to build. Why not this issue is not resolvable on Amplify's end?

**Comment by mikefrancis** [comment]
Spitballing temporary fixes just in case it helps...

Is there a reason to actually do anything to that pertains to SSR if nothing has changed?

I've just pushed 2 empty commits and they're both 10+ minutes build times. 

It's pretty crude but would something that simply checks if there's no diff in src/pages/api/**/* in the commit, skip the SSR build entirely?

**Comment by kantorcodes** [comment]
Hello, 

Any updates on these long build times?

**Comment by jeremyrajan** [comment]
+1, its takes more than 20mins for a medium-sized app to build. Compared to vercel or netlify.

**Comment by chernetsov** [comment]
@calavera +1 here, slow build times are super annoying, especially as you are just getting started

**Comment by Samia2Aimas** [comment]
Any updates about the feature ?

**Comment by OskarD** [comment]
I migrated my webapp to Vercel because of this so stay incommunicado for all I care, but if you want to keep other customers you should probably get working on this soon

**Comment by ctjlewis** [comment]
Amazon has the best engineers in the world full-time and we can't run a build script in less than 40 minutes. Entirely noncompetitive with every other deploy process (Vercel, Netlify, etc.).

**Comment by matart15** [comment]
Amplify is just too false advertised product.

It is only good for `Todo` app. Nothing more. 

It breaks often. Not just because of your code. Because of infra changes. And that case you can't do anything. You just pull your hairs with broken app. Few days later it magically works. ( this happened to me quiet a lot of times ). 

No one does not use it for bigger projects. Even amazon does not. I found few issue, people asking wheter amazon use Amplify on any product except amplify website. Amplify team just dodge the question.

In short don't use Amplify for your own future sake.

**Comment by entest-hai** [comment]
same experience, the same code but the first time failed, then next time works (just click deploy again). It is weird that it happens when I migrating to vercel.

**Comment by alexhhn** [comment]
I have 2 simple SSR pages, the build takes 12 min + each time, same issue here

**Comment by samparmenter** [comment]
I moved to vercel in the end. Amplify is a joke.

**Comment by khanakia** [comment]
I also faced the issue with amplify app my deployment stuck for 2 days and i tried to reupload the code but nothing worked. So I moved to EC2 it is rather easy to manage and i wasted a lot of time making the amplify work. It's a useless product.

**Comment by jeddtobee023** [comment]
Any updates on this?

Is moving to EC2 gonna be our best option for now?

**Comment by ganesh-rao** [comment]
The build process is still far too slow.

Why am I paying for build time to upgrade the Amplify cli on the machine that's cloning the repository?
![Capture1](https://user-images.githubusercontent.com/16326735/190988114-46b5c6ea-e851-4423-8f8c-6f5dc97a5012.JPG)

And perhaps something that could help speed things up?
![Capture2](https://user-images.githubusercontent.com/16326735/190988304-b35b78c7-461e-4ee5-a017-a467c5707517.JPG)
https://nextjs.org/docs/messages/no-cache
https://nextjs.org/docs/advanced-features/ci-build-caching#aws-codebuild

**Comment by mossen** [comment]
Is there any update soon, or all of us should move out?

**Comment by justyn-clark** [comment]
> Is there any update soon, or all of us should move out?

Move bro. I've been getting notifications for this thread for what feels like a year now.

I'm deploying on Fly.io and really happy with it.

**Comment by phoenixbeats01** [comment]
@justyn-clark oh ok let me just move my entire cloud infrastructure to another service just because of an annoying inconvenience.

**Comment by justyn-clark** [comment]
> @justyn-clark oh ok let me just move my entire cloud infrastructure to another service just because of an annoying inconvenience.

Exactly. Ever heard of Docker images?

**Comment by mossen** [comment]
My Next.js app is a static app by having the following changes I could make the build time to go down from ~15min to ~5min. Though still there is a huge different with Netlify/Vercel build time ~2min, it is way better than before.

- Not using `next/image` by either using `img` or disable the config. I guess with the static pages Vercel/Netlify doing some magic in the background which Amplify couldn't do that and that is why the build takes so long.
- Make the app to export the static files: `"build": "next build && next export"`
- update `amplify.yml` to use `out` directory: `baseDirectory: out`

**Comment by JohnDotOwl** [comment]
The build time is really long compared to vercel

**Comment by dinizgb** [comment]
Same here, awful service.

**Comment by nhtera** [comment]
Same issue. Amplify build is very slow compared to Vercel. The time building average is 10 minutes :(

**Comment by michaelworm** [comment]
NextJS 13 has been released, did someone test if the build times are faster now?

https://nextjs.org/blog/next-13#introducing-turbopack-alpha

**Comment by brkdmn35** [comment]
> NextJS 13 has been released, did someone test if the build times are faster now?
> 
> https://nextjs.org/blog/next-13#introducing-turbopack-alpha

Not available for production build for now. There is a comment under the conten,

Note: Turbopack in Next.js currently only supports next dev. View the [supported features](https://turbo.build/pack/docs/features). We are also working to add support for next build through Turbopack.

**Comment by ChristopherBanas** [comment]
Would any of this be helped by caching? I've been trying to implement a build cache but per https://github.com/aws-amplify/amplify-hosting/issues/2436 it does not seem like amplify supports this
